googlecloud-plugins/src/main/java/co/cask/hydrator/plugin/sink/GCSBatchSink.java

```diff
@@ -0,0 +1,152 @@
+
+  private final GCSSinkConfig config;
+
+  protected GCSBatchSink(GCSSinkConfig config) {
+    super(config);
+    this.config = config;
+    // update fileSystemProperties to include ProjectId and JsonKeyFile, so prepareRun can only set
+    // fileSystemProperties in configuration, and not deal with projectId and JsonKeyFile separately
+    // do not create file system properties if macros were provided unless in a test case
+    if (!this.config.containsMacro("fileSystemProperties") && !this.config.containsMacro("JsonKeyFile") &&
+      !this.config.containsMacro("ProjectId")) {
+      this.config.fileSystemProperties = updateFileSystemProperties(this.config.fileSystemProperties,
+                                                                    this.config.projectId, this.config.jsonKey);
+    }
+  }
+
+  @VisibleForTesting
+  GCSSinkConfig getConfig() {
+    return config;
+  }
+
+  @Override
+  public final void prepareRun(BatchSinkContext context) {
+    OutputFormatProvider outputFormatProvider = createOutputFormatProvider(context);
+    Map<String, String> outputConfig = new HashMap<>(outputFormatProvider.getOutputFormatConfiguration());
+    if (config.fileSystemProperties != null) {
+      Map<String, String> properties = GSON.fromJson(config.fileSystemProperties, MAP_STRING_STRING_TYPE);
+      outputConfig.putAll(properties);
+    }
+    context.addOutput(Output.of(config.referenceName, new SinkOutputFormatProvider(
+      outputFormatProvider.getOutputFormatClassName(), outputConfig)));
+  }
+
+  protected abstract OutputFormatProvider createOutputFormatProvider(BatchSinkContext context);
+
+  private static String updateFileSystemProperties(@Nullable String fileSystemProperties,
+                                                   String projectId, String jsonKey) {
+    Map<String, String> providedProperties;
+    if (fileSystemProperties == null) {
+      providedProperties = new HashMap<>();
+    } else {
+      providedProperties = GSON.fromJson(fileSystemProperties, MAP_STRING_STRING_TYPE);
+    }
+    providedProperties.put("fs.gs.project.id", projectId);
+    providedProperties.put("google.cloud.auth.service.account.json.keyfile", jsonKey);
+    return GSON.toJson(providedProperties);
+  }
+
+  /**
+   * GCS Sink configuration.
+   */
+  public static class GCSSinkConfig extends ReferencePluginConfig {
+
+    @Name("BucketKey")
+    @Description(BUCKET_DES)
+    @Macro
+    protected String bucketKey;
+
+    @Name("ProjectId")
+    @Description(PROJECT_ID_DES)
+    @Macro
+    protected String projectId;
+
+    @Name("JsonKeyFile")
+    @Description(SERVICE_KEY_FILE_DES)
+    @Macro
+    protected String jsonKey;
+
+    @Name("BucketDir")
+    @Description("the directory inside the bucket where the data is stored. Need to be a new directory.")
+    @Macro
+    protected String bucketDir;
+
+    @Description(FILESYSTEM_PROPERTIES_DESCRIPTION)
+    @Nullable
+    @Macro
+    protected String fileSystemProperties;
```

Comment:
Its better for this to be a list of key value pairs. There is already a 'keyvalue' widget for a list of key value pairs that is more user friendly than writing json.